File: lecture04/01_palindromic/palindromic.py
#!/usr/bin/env python3

import collections
import sys

# Functions

def is_palindromic(s):
    ''' Counting Solution '''

    # counts = collections.Counter(s)
    counts = {}
    for c in s:
        counts[c] = counts.get(c, 0) + 1

    odds = 0
    for count in counts.values():
        odds += count % 2

    return odds <= 1

# Main Execution

if __name__ == '__main__':
    for word in sys.stdin:
        print('Yes' if is_palindromic(word.rstrip()) else 'No')

File: lecture04/01_palindromic/palindromic.cpp
// palindromic.cpp

#include <iostream>
#include <string>
#include <unordered_set>

using namespace std;

// Functions

bool is_palindromic(string &s) {
    unordered_set<char> seen;

    // History approach: keep track of whether or not we have seen this
    // character before.
    for (auto c : s) {
    	if (seen.count(c)) {
    	    seen.erase(c);
	} else {
	    seen.insert(c);
	}
    }

    return seen.size() <= 1;
}

// Main execution

int main(int argc, char *argv[]) {
    string word;

    while (cin >> word) {
    	cout << (is_palindromic(word) ? "Yes" : "No") << endl;
    }

    return 0;
}
